cirq/protocols/json.py

```diff
@@ -0,0 +1,187 @@
+            self._crd = {
+                'CCXPowGate': cirq.CCXPowGate,
+                'CCZPowGate': cirq.CCZPowGate,
+                'CNotPowGate': cirq.CNotPowGate,
+                'CSwapGate': cirq.CSwapGate,
+                'CZPowGate': cirq.CZPowGate,
+                'Circuit': cirq.Circuit,
+                'FSimGate': cirq.FSimGate,
+                'GateOperation': cirq.GateOperation,
+                'GlobalPhaseOperation': cirq.GlobalPhaseOperation,
+                'GridQubit': cirq.GridQubit,
+                'HPowGate': cirq.HPowGate,
+                'ISwapPowGate': cirq.ISwapPowGate,
+                'IdentityGate': cirq.IdentityGate,
+                'LineQubit': cirq.LineQubit,
+                'MeasurementGate': cirq.MeasurementGate,
+                'Moment': cirq.Moment,
+                'NamedQubit': cirq.NamedQubit,
+                '_PauliX': cirq.ops.pauli_gates._PauliX,
+                '_PauliY': cirq.ops.pauli_gates._PauliY,
+                '_PauliZ': cirq.ops.pauli_gates._PauliZ,
+                'PauliString': cirq.PauliString,
+                'PhasedXPowGate': cirq.PhasedXPowGate,
+                'SingleQubitPauliStringGateOperation':
+                cirq.SingleQubitPauliStringGateOperation,
+                'SwapPowGate': cirq.SwapPowGate,
+                '_UnconstrainedDeviceType':
+                cirq.devices.unconstrained_device._UnconstrainedDeviceType,
+                'XPowGate': cirq.XPowGate,
+                'XXPowGate': cirq.XXPowGate,
+                'YPowGate': cirq.YPowGate,
+                'YYPowGate': cirq.YYPowGate,
+                'ZPowGate': cirq.ZPowGate,
+                'ZZPowGate': cirq.ZZPowGate,
+
+                # not a cirq class, but treated as one:
+                'complex': complex,
+            }
+        return self._crd
+
+
+RESOLVER_CACHE = _ResolverCache()
+
+
+def _cirq_class_resolver(cirq_type: str) -> Union[None, Type]:
+    return RESOLVER_CACHE.cirq_class_resolver_dictionary.get(cirq_type, None)
+
+
+DEFAULT_RESOLVERS = [
+    _cirq_class_resolver,
+]
+
+
+class SupportsJSON(Protocol):
+    """An object that can be turned into JSON dictionaries.
+
+    The magic method _json_dict_ must return a trivially json-serializable
+    type or other objects that support the SupportsJSON protocol.
+
+    During deserialization, a class must be able to be resolved (see
+    the docstring for `read_json`) and must be able to be (re-)constructed
+    from the serialized parameters. If the type defines a classmethod
+    `_from_json_dict_`, that will be called. Otherwise, the `cirq_type` key
+    will be popped from the dictionary and used as kwargs to the type's
+    constructor.
+    """
+
+    def _json_dict_(self) -> Union[None, NotImplementedType, Dict[Any, Any]]:
+        pass
+
+
+def to_json_dict(obj, attribute_names):
+    d = {'cirq_type': obj.__class__.__name__}
+    for attr_name in attribute_names:
```

Comment:
I guess we want more control than just doing introspection on the attributes passed to the constructor? 

One thought is that you could use annotations to control this:
`def __init__(exponent: SerializableFloat):`

Reply:
> doing introspection on the attributes passed to the constructor

This is a decent idea, but I'm worried it veers too far towards "magic". The current scheme makes it pretty easy to add JSON support for a class, but a developer must explicitly add something to indicate it supports JSON. If this part was automated there is a chance that a class will serialize badly. Right now, if you try to serialize something that a developer hasn't given a little bit of thought to, it will just refuse. I think that's a desirable thing

Reply:
The problem with the annotations in the constructor is it would introduce noise to users who might be entirely uninterested in serialization

Reply:
That's true.  It's a very Java way to handle this problem :)